src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -174,22 +183,57 @@ private void GenerateProlog()
                 signatureIndex++;
             }
 
+            string[] argNames = null;
+            if (_debugInformation != null)
+            {
+                argNames = _debugInformation.GetParameterNames()?.ToArray();
+            }
+
             for (int i = 0; i < _signature.Length; i++)
             {
                 if (CanStoreTypeOnStack(_signature[i]))
                 {
-                    LLVMValueRef argStackSlot = LLVM.BuildAlloca(_builder, GetLLVMTypeForTypeDesc(_signature[i]), $"arg{i + thisOffset}_");
+                    string argName = String.Empty;
+                    if (argNames != null && argNames[i] != null)
+                    {
+                        argName = argNames[i] + "_";
+                    }
+                    argName += $"arg{i + thisOffset}_";
+
+                    LLVMValueRef argStackSlot = LLVM.BuildAlloca(_builder, GetLLVMTypeForTypeDesc(_signature[i]), argName);
                     LLVM.BuildStore(_builder, LLVM.GetParam(_llvmFunction, (uint)signatureIndex), argStackSlot);
                     _argSlots[i] = argStackSlot;
                     signatureIndex++;
                 }
             }
 
+            string[] localNames = new string[_locals.Length];
+            if (_debugInformation != null)
+            {
+                foreach (ILLocalVariable localDebugInfo in _debugInformation.GetLocalVariables() ?? Enumerable.Empty<ILLocalVariable>())
+                {
+                    // Check whether the slot still exists as the compiler may remove it for intrinsics
```

Comment:
Was this part needed because of the replaced method bodies? It might not be necessary anymore.

I tend to prefer the compiler crashing instead of silently ignoring bad states (and potentially generating garbage that is harder to reason about as a result).

Reply:
Yes it was originally. I'm not sure if it still is after switching to the other way of getting DebugInfo